airflow/models/param.py

```
@@ -0,0 +1,208 @@
+
+from airflow.exceptions import AirflowException
+
+
+class Param:
+    """
+    Class to hold the default value of a Param and rule set to do the validations. Without the rule set
+    it always validates and returns the default value.
+
+    :param default: The value of this Param object holds
+    :type default: Any
+    :param description: Optional help text for the Param
+    :type description: str
+    :param schema: The validation schema of the Param, if not given then all kwargs except
+        default & description will form the schema
+    :type schema: dict
+    """
+
+    def __init__(self, default: Any = None, description: str = None, **kwargs):
+        self.default = default
+        self.description = description
+        self.schema = kwargs.pop('schema') if 'schema' in kwargs else kwargs
+
+        # If default is not None, then validate it once, may raise ValueError
+        if default:
+            try:
+                jsonschema.validate(self.default, self.schema, format_checker=FormatChecker())
+            except ValidationError as err:
+                raise ValueError(err)
+
+    def resolve(self, value: Optional[Any] = None, suppress_exception: bool = False) -> Any:
+        """
+        Runs the validations and returns the Param's final value.
+        May raise ValueError on failed validations.
+
+        :param value: The value to be updated for the Param
+        :type: Optional[Any]
+        :param suppress_exception: To raise an exception or not when the validations fails.
+            If true and validations fails, the return value would be None.
+        :type suppress_exception: bool
+        """
+        try:
+            final_val = value or self.default
+            jsonschema.validate(final_val, self.schema, format_checker=FormatChecker())
+            self.default = final_val
+        except ValidationError as err:
+            if suppress_exception:
+                return None
+            raise ValueError(err) from None
```

If we are refactoring this code maybe it would be good to raise `ValidationError` instead of `ValueError`. I think it would be more appropriate. Also, do we need this `from None`?

---

To suppress the exception context chain, see https://www.python.org/dev/peps/pep-0409

---

Oh nice, I didn't know about that!